Problem: upgrading to tendermint-* 0.25 crates fail

```
error[E0308]: mismatched types
[359](https://github.com/crypto-com/defi-wallet-core-rs/actions/runs/3177186515/jobs/5177346363#step:5:360)
  --> bindings/wasm/src/cosmos_sdk.rs:74:20
[360](https://github.com/crypto-com/defi-wallet-core-rs/actions/runs/3177186515/jobs/5177346363#step:5:361)
   |
[361](https://github.com/crypto-com/defi-wallet-core-rs/actions/runs/3177186515/jobs/5177346363#step:5:362)
74 |             if let tendermint::abci::Code::Err(_) = resp.code {
[362](https://github.com/crypto-com/defi-wallet-core-rs/actions/runs/3177186515/jobs/5177346363#step:5:363)
   |                    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^   --------- this expression has type `tendermint::abci::code::Code`
[363](https://github.com/crypto-com/defi-wallet-core-rs/actions/runs/3177186515/jobs/5177346363#step:5:364)
   |                    |
[364](https://github.com/crypto-com/defi-wallet-core-rs/actions/runs/3177186515/jobs/5177346363#step:5:365)
   |                    expected enum `tendermint::abci::code::Code`, found enum `tendermint::abci::Code`
[365](https://github.com/crypto-com/defi-wallet-core-rs/actions/runs/3177186515/jobs/5177346363#step:5:366)
   |
[366](https://github.com/crypto-com/defi-wallet-core-rs/actions/runs/3177186515/jobs/5177346363#step:5:367)
   = note: perhaps two different versions of crate `tendermint` are being used?
[367](https://github.com/crypto-com/defi-wallet-core-rs/actions/runs/3177186515/jobs/5177346363#step:5:368)
```

--- a/.github/workflows/audit.yml
+++ b/.github/workflows/audit.yml
@@ -20,4 +20,5 @@ jobs:
       - uses: dtolnay/rust-toolchain@stable
       - name: Install cargo audit
         run: cargo install cargo-audit
-      - run: cargo audit --deny warnings
+        # RUSTSEC-2021-0145: `atty` transitive dependency, only informational for Windows (not exploitable in practice)
+      - run: cargo audit --deny warnings --ignore RUSTSEC-2021-0145

--- a/bindings/cpp/Cargo.toml
+++ b/bindings/cpp/Cargo.toml
@@ -8,7 +8,7 @@ license = "Apache-2.0"
 [dependencies]
 defi-wallet-core-common = { path = "../../common" , features=["login"]}
 defi-wallet-core-proto = { version = "0.1", path = "../../proto" }
-cosmos-sdk-proto = { version = "0.14" }
+cosmos-sdk-proto = { version = "0.15" }
 cxx = "1"
 anyhow = "1"
 serde="1"

--- a/bindings/wasm/Cargo.toml
+++ b/bindings/wasm/Cargo.toml
@@ -29,14 +29,14 @@ wasm-bindgen = "0.2"
 # all the `std::fmt` and `std::panicking` infrastructure, so isn't great for
 # code size when deploying.
 console_error_panic_hook = { version = "0.1", optional = true }
-cosmos-sdk-proto = { version = "0.14", default-features = false, features = ["cosmwasm"] }
+cosmos-sdk-proto = { version = "0.15", default-features = false, features = ["cosmwasm"] }
 
-tendermint = "0.23"
+tendermint = "0.26"
 
 [dev-dependencies]
 wasm-bindgen-test = "0.3"
 ethers = { version = "0.17", features = ["rustls"] }
 wasm-timer = "0.2"
-tendermint-rpc = "0.23"
+tendermint-rpc = "0.26"
 defi-wallet-core-proto = { version = "0.1", path = "../../proto" }
 tonic-web-wasm-client = "0.2"

--- a/bindings/wasm/src/cosmos_sdk.rs
+++ b/bindings/wasm/src/cosmos_sdk.rs
@@ -55,7 +55,7 @@ impl CosmosClient {
                 .into_result()
                 .map_err(format_to_js_error)?;
 
-            if let tendermint::abci::Code::Err(_) = resp.code {
+            if resp.code.is_err() {
                 return Err(serde_wasm_bindgen::to_value(&resp).map_err(format_to_js_error)?);
             }
 

--- a/bindings/wasm/tests/nft.rs
+++ b/bindings/wasm/tests/nft.rs
@@ -157,7 +157,7 @@ impl NftClient {
             .into_serde::<Response>()
             .unwrap();
 
-        assert_eq!(res.code, tendermint::abci::Code::Ok);
+        assert!(res.code.is_ok());
         wait_for_timeout(None).await;
     }
 }
@@ -189,31 +189,31 @@ async fn validate_issue() {
         .supply(denom1.id.clone(), SIGNER1.to_owned())
         .await;
     assert!(res.is_ok());
-    let supply = res.unwrap().into_serde::<u64>().unwrap();
+    let supply = serde_wasm_bindgen::from_value::<u64>(res.unwrap()).unwrap();
     assert_eq!(supply, 0);
 
     let res = grpc_client
         .owner(denom1.id.clone(), SIGNER1.to_owned(), None)
         .await;
     assert!(res.is_ok());
-    let owner = res.unwrap().into_serde::<Owner>().unwrap();
+    let owner = serde_wasm_bindgen::from_value::<Owner>(res.unwrap()).unwrap();
     assert_eq!(owner.address, SIGNER1.to_owned());
     assert_eq!(owner.id_collections, vec![]);
 
     let res = grpc_client.collection(denom2.id.clone(), None).await;
     assert!(res.is_ok());
-    let collection = res.unwrap().into_serde::<Collection>().unwrap();
+    let collection = serde_wasm_bindgen::from_value::<Collection>(res.unwrap()).unwrap();
     assert_eq!(collection.denom, Some(denom2.clone()));
     assert_eq!(collection.nfts.len(), 0);
 
     let res = grpc_client.denom(denom2.id.clone()).await;
     assert!(res.is_ok());
-    let denom = res.unwrap().into_serde::<Denom>().unwrap();
+    let denom = serde_wasm_bindgen::from_value::<Denom>(res.unwrap()).unwrap();
     assert_eq!(denom, denom2);
 
     let res = grpc_client.denom_by_name(denom1.name.clone()).await;
     assert!(res.is_ok());
-    let denom = res.unwrap().into_serde::<Denom>().unwrap();
+    let denom = serde_wasm_bindgen::from_value::<Denom>(res.unwrap()).unwrap();
     assert_eq!(denom, denom1);
 
     let pagination = PageRequest {
@@ -225,7 +225,7 @@ async fn validate_issue() {
     };
     let res = grpc_client.denoms(Some(pagination)).await;
     assert!(res.is_ok());
-    let denoms = res.unwrap().into_serde::<Vec<Denom>>().unwrap();
+    let denoms = serde_wasm_bindgen::from_value::<Vec<Denom>>(res.unwrap()).unwrap();
     assert_eq!(denoms.len(), 2);
     assert_eq!(denoms[0], denom2);
     assert_eq!(denoms[1], denom1);
@@ -240,13 +240,13 @@ async fn validate_mint() {
         .nft(denom1.id.clone(), TEST_TOKEN_ID.to_owned())
         .await;
     assert!(res.is_ok());
-    let nft = res.unwrap().into_serde::<BaseNft>().unwrap();
+    let nft = serde_wasm_bindgen::from_value::<BaseNft>(res.unwrap()).unwrap();
     assert_eq!(nft.owner, SIGNER2.to_owned());
 
     // Check collection after minting.
     let res = grpc_client.collection(denom1.id.clone(), None).await;
     assert!(res.is_ok());
-    let collection = res.unwrap().into_serde::<Collection>().unwrap();
+    let collection = serde_wasm_bindgen::from_value::<Collection>(res.unwrap()).unwrap();
     assert_eq!(collection.denom, Some(denom1));
     assert_eq!(collection.nfts.len(), 1);
 }
@@ -260,7 +260,7 @@ async fn validate_before_transfer() {
         .owner(denom1.clone().id, SIGNER2.to_owned(), None)
         .await;
     assert!(res.is_ok());
-    let owner = res.unwrap().into_serde::<Owner>().unwrap();
+    let owner = serde_wasm_bindgen::from_value::<Owner>(res.unwrap()).unwrap();
     assert_eq!(owner.address, SIGNER2.to_owned());
     assert_eq!(owner.id_collections.len(), 1);
     assert_eq!(owner.id_collections[0].denom_id, denom1.clone().id);
@@ -279,7 +279,7 @@ async fn validate_after_transfer() {
         .nft(denom1.clone().id, TEST_TOKEN_ID.to_owned())
         .await;
     assert!(res.is_ok());
-    let nft = res.unwrap().into_serde::<BaseNft>().unwrap();
+    let nft = serde_wasm_bindgen::from_value::<BaseNft>(res.unwrap()).unwrap();
     assert_eq!(nft.owner, SIGNER1.to_owned());
 
     // Check owner.
@@ -287,7 +287,7 @@ async fn validate_after_transfer() {
         .owner(denom1.clone().id, SIGNER2.to_owned(), None)
         .await;
     assert!(res.is_ok());
-    let owner = res.unwrap().into_serde::<Owner>().unwrap();
+    let owner = serde_wasm_bindgen::from_value::<Owner>(res.unwrap()).unwrap();
     assert_eq!(owner.address, SIGNER2.to_owned());
     assert_eq!(owner.id_collections, vec![]);
 }
@@ -300,7 +300,7 @@ async fn validate_edit() {
         .nft(denom1.id.clone(), TEST_TOKEN_ID.to_owned())
         .await;
     assert!(res.is_ok());
-    let nft = res.unwrap().into_serde::<BaseNft>().unwrap();
+    let nft = serde_wasm_bindgen::from_value::<BaseNft>(res.unwrap()).unwrap();
     assert_eq!(nft.name, "newname".to_owned());
     assert_eq!(nft.uri, "newuri".to_owned());
     assert_eq!(nft.data, "newdata".to_owned());
@@ -315,7 +315,7 @@ async fn validate_before_burn() {
         .supply(denom1.id.clone(), SIGNER1.to_owned())
         .await;
     assert!(res.is_ok());
-    let supply = res.unwrap().into_serde::<u64>().unwrap();
+    let supply = serde_wasm_bindgen::from_value::<u64>(res.unwrap()).unwrap();
     assert_eq!(supply, 1);
 }
 
@@ -333,13 +333,13 @@ async fn validate_after_burn() {
         .supply(denom1.clone().id, SIGNER1.to_owned())
         .await;
     assert!(res.is_ok());
-    let supply = res.unwrap().into_serde::<u64>().unwrap();
+    let supply = serde_wasm_bindgen::from_value::<u64>(res.unwrap()).unwrap();
     assert_eq!(supply, 0);
 
     // Check collection.
     let res = grpc_client.collection(denom1.clone().id, None).await;
     assert!(res.is_ok());
-    let collection = res.unwrap().into_serde::<Collection>().unwrap();
+    let collection = serde_wasm_bindgen::from_value::<Collection>(res.unwrap()).unwrap();
     assert_eq!(collection.denom, Some(denom1));
     assert_eq!(collection.nfts.len(), 0);
 }

--- a/bindings/wasm/tests/test_helper.rs
+++ b/bindings/wasm/tests/test_helper.rs
@@ -97,12 +97,12 @@ pub(crate) fn tonic_web_wasm_client() -> GrpcWebClient {
 }
 
 pub(crate) async fn query_chainmain_account(address: &str) -> RawRpcAccountStatus {
-    let account_details =
+    let account_details = serde_wasm_bindgen::from_value::<RawRpcAccountResponse>(
         JsFuture::from(chainmain_client().query_account_details(address.to_owned()))
             .await
-            .unwrap()
-            .into_serde::<RawRpcAccountResponse>()
-            .unwrap();
+            .unwrap(),
+    )
+    .unwrap();
 
     match account_details {
         RawRpcAccountResponse::OkResponse { account } => account,

--- a/common/Cargo.toml
+++ b/common/Cargo.toml
@@ -23,11 +23,11 @@ anyhow = "1"
 base64 = "0.13"
 bech32 = "0.9"
 bip39 = "1"
-cosmrs = "0.9"
+cosmrs = "0.10"
 eyre = "0.6"
 ethers = { version = "0.17", features = ["rustls", "abigen"] }
-ibc = { version = "0.19", default-features = false }
-ibc-proto = { version = "0.20", default-features = false }
+ibc = { version = "0.23", default-features = false }
+ibc-proto = { version = "0.22", default-features = false }
 itertools = "0.10"
 lazy_static = "1"
 pest = { version = "2", optional = true }
@@ -42,16 +42,16 @@ serde = "1"
 serde_json = "1"
 serde_with = "2"
 siwe = { version = "0.5", optional = true }
-tendermint = "0.23"
-tendermint-proto = "0.23"
-tendermint-rpc = "0.23"
+tendermint = "0.26"
+tendermint-proto = "0.26"
+tendermint-rpc = "0.26"
 thiserror = "1"
 uniffi = { version = "^0.21", optional = true }
 uniffi_macros = { version = "^0.21", optional = true }
 url = "2"
 
 [target.'cfg(target_arch = "wasm32")'.dependencies]
-cosmos-sdk-proto = { version = "0.14", default-features = false, features = ["cosmwasm", "grpc"] }
+cosmos-sdk-proto = { version = "0.15", default-features = false, features = ["cosmwasm", "grpc"] }
 defi-wallet-core-proto = { version = "0.1", path = "../proto" }
 # NOTE: crate `bip39` cannot work with latest crate `rand` (0.8.0)
 # FIXME: https://github.com/rust-bitcoin/rust-bip39/issues/14
@@ -63,7 +63,7 @@ wasm-bindgen = { version = "0.2", features = ["serde-serialize"] }
 tonic-web-wasm-client = "0.2"
 
 [target.'cfg(not(target_arch = "wasm32"))'.dependencies]
-cosmos-sdk-proto = { version = "0.14", features = ["grpc"] }
+cosmos-sdk-proto = { version = "0.15", features = ["grpc"] }
 defi-wallet-core-proto = { version = "0.1", path = "../proto", features = ["transport"] }
 # NOTE: crate `bip39` cannot work with latest crate `rand` (0.8.0)
 # FIXME: https://github.com/rust-bitcoin/rust-bip39/issues/14

--- a/common/src/transaction/cosmos_sdk/parser/base_parser.rs
+++ b/common/src/transaction/cosmos_sdk/parser/base_parser.rs
@@ -9,7 +9,7 @@ use cosmos_sdk_proto::cosmos::staking::v1beta1::{MsgBeginRedelegate, MsgDelegate
 use cosmos_sdk_proto::traits::{Message, TypeUrl};
 use eyre::WrapErr;
 use ibc::applications::transfer::msgs::transfer;
-use tendermint_proto::Protobuf;
+use ibc_proto::protobuf::Protobuf;
 
 /// Base parser for standard Cosmos messages
 pub struct BaseParser;

--- a/common/src/transaction/cosmos_sdk/parser/structs.rs
+++ b/common/src/transaction/cosmos_sdk/parser/structs.rs
@@ -85,7 +85,7 @@ impl From<Fee> for CosmosFee {
 
         Self {
             amount,
-            gas_limit: fee.gas_limit.value(),
+            gas_limit: fee.gas_limit,
             payer: fee.payer.map(|p| p.to_string()),
             granter: fee.granter.map(|g| g.to_string()),
         }

--- a/common/src/transaction/nft.rs
+++ b/common/src/transaction/nft.rs
@@ -736,7 +736,7 @@ mod test {
         let chain_id = "chaintest".parse()?;
         let account_number = 1;
         let sequence_number = 0;
-        let gas = 100_000;
+        let gas = 100_000u64;
         let timeout_height = 9001u16;
         let memo = "example memo";
 

--- a/proto/Cargo.toml
+++ b/proto/Cargo.toml
@@ -11,11 +11,11 @@ edition = "2021"
 rust-version = "1.57"
 
 [dependencies]
-cosmrs = "0.9"
-cosmos-sdk-proto = { version = "0.14", default-features = false }
+cosmrs = "0.10"
+cosmos-sdk-proto = { version = "0.15", default-features = false }
 prost = "0.11"
 prost-types = "0.11"
-tendermint-proto = "0.23"
+tendermint-proto = "0.26"
 tonic = { version = "0.8", default-features = false, features = ["codegen", "prost"] }
 serde = "1"
 

